### DISCUSSION TITLE
How to process the data before sending it?

### DISCUSSION THREAD

**Discussion by DrakoPOD** [moscajs/aedes#627] [Q&A] [answered]
Hello, I need to pack the data from the MQTT in a JSON before send to the browser, devices are sending data every 2ms, but my webapp turns too laggy when I process this data. So, I'm trying to send every 200ms in an array or JSON. Is there a way to do that?

**Comment by robertsLando** [comment, marked as answer]
Could you explain better please? This seems not a problem with aedes itself as aedes is just the broker

**Reply by DrakoPOD** [reply]
Of course, is not a problem, this is Q&A page. My question is how to do this: I need to release work on the client side.

I have inefficient solution: a device send data through broker, then another MQTT client receive it, process it and resend it to the broker with a different topic, then the web client receives the processed data.

I want to know if is there a way that the broker processes the data. At the moment, there isn't much info about Aedes on internet, even docs are lacking of examples. Maybe someone know how to help me.

**Reply by robertsLando** [reply]
You can use this: https://github.com/moscajs/aedes/blob/main/docs/Aedes.md#handler-authorizepublish-client-packet-callback

**Reply by DrakoPOD** [reply]
Well, I tried, but only worked the first time, and it stop sending data
```javascript
aedes.authorizePublish = function (client, packet, callback) {
  //t[0] is the device name, t[1] is the type of data
  let t = packet.topic.split("/");

  //Get device specs
  if (t[1] === 'params') {
    const pack = JSON.parse(packet.payload)
    devicesData[pack.name] = {}
    for (let i = 0; i < pack.sensor.length; i++) {
      devicesData[t[0]][pack.sensor[i]] = []
    }
    return callback(null)
  }

  if (t[0] in devicesData) {
    //Save data as array
    devicesData[t[0]][t[1]].push(packet.payload)
    console.log(devicesData[t[0]][t[1]].length)

    //Send data when is reach the limit and restart
    if(devicesData[t[0]][t[1]].length >= 10){
      packet.payload = devicesData[t[0]][t[1]]
      devicesData[t[0]][t[1]] = []

      return callback(null)
    }
    return;
  }
}
```

**Reply by robertsLando** [reply]
You must always call the callback. There is a return statement at the end without a callback

**Reply by DrakoPOD** [reply]
Yes, I know, but if I do that the packet will be publish, and I want to retain it, callback only have two states, `null` and `error`, I don't want to return an error to the device.

**Reply by robertsLando** [reply]
Without a callback that will never work. Return null or an error

**Reply by DrakoPOD** [reply]
Well, I just redirected the packet to an unused topic, but how I can delete or prevent the packet send without return an error? Is just for have cleanest code.

**Reply by robertsLando** [reply]
Returning an error is the only way to reject a publish

**Reply by DrakoPOD** [reply]
Oh!, you should add a way to destroy a packet, could be a problem if the MQTT client have qos 1. Also, the server doesn't need to process a data that doesn't go anywhere.
